modules/Movecraft/src/main/java/net/countercraft/movecraft/async/rotation/RotationTask.java

```diff
@@ -377,6 +341,82 @@ public boolean getIsSubCraft() {
         return isSubCraft;
     }
 
+    private boolean checkFuel(){
+        // check for fuel, burn some from a furnace if needed. Blocks of coal are supported, in addition to coal and charcoal
+        double fuelBurnRate = craft.getType().getFuelBurnRate(craft.getW());
+
+        if (fuelBurnRate == 0.0 || craft.getSinking()) {
+            return true;
+        }
+        if (craft.getBurningFuel() >= fuelBurnRate) {
+            double burningFuel = craft.getBurningFuel();
+            //call event
+            final FuelBurnEvent event = new FuelBurnEvent(craft, burningFuel, fuelBurnRate);
+            Bukkit.getPluginManager().callEvent(event);
+            if (event.getBurningFuel() != burningFuel)
+                burningFuel = event.getBurningFuel();
+            if (event.getFuelBurnRate() != fuelBurnRate)
+                fuelBurnRate = event.getFuelBurnRate();
+            craft.setBurningFuel(burningFuel - fuelBurnRate);
+            return true;
+        }
+        Block fuelHolder = null;
+        for (MovecraftLocation bTest : oldHitBox) {
+            Block b = craft.getW().getBlockAt(bTest.getX(), bTest.getY(), bTest.getZ());
+            if (b.getType() == Material.FURNACE) {
+                InventoryHolder inventoryHolder = (InventoryHolder) b.getState();
+                for (Material fuel : craft.getType().getFuelTypes().keySet()) {
+                    if (!inventoryHolder.getInventory().contains(fuel))
```

Comment:
loop over the inventory rather than the keyset, as the inventory should always be size 3 while the keyset can scale.